tests/unit/test_charm.py

```diff
@@ -1753,3 +1756,66 @@ def test_given_vault_available_when_vault_pki_certificate_creation_request_then_
             ca=ca,
             chain=chain,
         )
+
+    @patch("charm.get_common_name_from_certificate", new=Mock)
+    @patch(f"{TLS_CERTIFICATES_LIB_PATH}.TLSCertificatesRequiresV3.get_assigned_certificates")
+    def test_given_vault_is_available_when_pki_certificate_is_available_then_default_issuer_is_set_to_latest_issuer(
+        self,
+        patch_get_assigned_certificates,
+    ):
+        self.mock_vault.configure_mock(
+            **{
+                "is_initialized.return_value": True,
+                "is_api_available.return_value": True,
+                "is_pki_role_created.return_value": False,
+                "get_intermediate_ca.return_value": "vault",
+                "is_common_name_allowed_in_pki_role.return_value": False,
+            },
+        )
+        csr = "some csr content"
+        certificate = "some certificate"
+        ca = "some ca"
+        chain = [ca]
+        self.harness.update_config({"common_name": "vault"})
+        self.harness.set_leader(is_leader=True)
+        self.harness.set_can_connect(container=self.container_name, val=True)
+        peer_relation_id = self._set_peer_relation()
+        self._set_approle_secret(
+            role_id="root token content",
+            secret_id="whatever secret id",
+        )
+        self._set_csr_secret_in_peer_relation(relation_id=peer_relation_id, csr="some csr content")
+        event = CertificateAvailableEvent(
+            handle=Mock(),
+            certificate=certificate,
+            certificate_signing_request=csr,
+            ca=ca,
+            chain=chain,
+        )
+        relation_id = self.harness.add_relation(
+            relation_name=TLS_CERTIFICATES_PKI_RELATION_NAME, remote_app="tls-provider"
+        )
+
+        patch_get_assigned_certificates.return_value = [
+            ProviderCertificate(
+                relation_id=relation_id,
+                application_name="tls-provider",
+                csr=csr,
+                certificate=certificate,
+                ca=ca,
+                chain=chain,
+                revoked=False,
+                expiry_time=datetime.now(timezone.utc),
+            )
+        ]
+
+        self.harness.charm._on_tls_certificate_pki_certificate_available(event=event)
+
+        self.mock_vault.set_pki_intermediate_ca_certificate.assert_called_with(
+            certificate=certificate,
+            mount=PKI_MOUNT,
+        )
+        self.mock_vault.create_or_update_pki_charm_role.assert_called_with(
+            allowed_domains="vault", mount=PKI_MOUNT, role=PKI_ROLE_NAME
+        )
+        self.mock_vault.make_latest_pki_issuer_default.assert_called_with(mount=PKI_MOUNT)
```

Comment:
Is there a more concrete way we could test this? This seems a bit to indirect and not really validating that it has the right issuer, and as this is a temporary fix, the next one will probably require changing this test again.

Reply:
Whether this is a temporary change or not depends on design decisions on how to handle CA changes not on implementation decisions, is it a problem in that case that the tests might change?

As for testing this in a more concrete way I added tests for the `make_latest_pki_issuer_default` in the client tests.
I don't know if that is enough but here in the charm tests we don't seem to validate that changes made by the client really took effect. For example we don't test that the role desired by `create_or_update_pki_charm_role` has been actually created. For us to know which issuer is being used we need to read the config on `/pki/config/issuers`

Anyway I do see your point and I agree with it, but I don't think it is unique to this function.

Reply:
How about an integration test step [here](https://github.com/canonical/vault-k8s-operator/blob/4e412beb4ceb1aa73aebd7e10b30464b47fc7fa0/tests/integration/test_integration.py#L386)? just add a change config in between and validate with a new requirer?

Reply:
Good point @kayra1 
I added integration tests.